src/resources/wifi_esp32.cc

```diff
@@ -305,23 +337,47 @@ PRIMITIVE(connect) {
   return proxy;
 }
 
+PRIMITIVE(establish) {
+  ARGS(WifiResourceGroup, group, cstring, ssid, cstring, password, bool, broadcast, int, channel);
+  HeapTagScope scope(ITERATE_CUSTOM_TAGS + WIFI_MALLOC_TAG);
+
+  if (ssid == null || password == null) {
+    INVALID_ARGUMENT;
+  }
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  WifiEvents* wifi = _new WifiEvents(group);
+  if (wifi == null) MALLOC_FAILED;
+
+  group->register_resource(wifi);
+
+  esp_err_t err = group->establish(ssid, password, broadcast, channel);
+  if (err != ESP_OK) {
+    group->unregister_resource(wifi);
+    return Primitive::os_error(err, process);
```

Comment:
Doesn't this leak wifi?

Reply:
No. Unregistering the resource deletes it.